File: __LEGACY__/ds-leetcode/russian-doll-envelopes.java
class Solution {
    public int maxEnvelopes(int[][] envelopes) {
        Arrays.sort(envelopes, (a, b) -> a[0] == b[0] ? b[1] - a[1] : a[0] - b[0]);
        int[] dp = new int[envelopes.length];
        int res = 0;
        for (int[] envelope : envelopes) {
            int i = Arrays.binarySearch(dp, 0, res, envelope[1]);
            if (i < 0) {
                i = -(i + 1);
            }
            dp[i] = envelope[1];
            if (i == res) {
                res++;
            }
        }
        return res;
    }
}

File: __LEGACY__/ds-leetcode/russian-doll-envelopes.js
/**
 * @param {number[][]} envelopes
 * @return {number}
 */
var maxEnvelopes = function(envelopes) {
    // Dynamic Programming
    // O(n)

    // Sort by width
    envelopes.sort((a, b) => a[0] - b[0]);

    // Sort by height
    envelopes.sort((a, b) => a[1] - b[1]);

    // Dynamic Programming
    // O(n)
    let dp = new Array(envelopes.length).fill(0);
    dp[0] = 1;
    for (let i = 1; i < envelopes.length; i++) {
        let max = 0;
        for (let j = 0; j < i; j++) {
            if (envelopes[i][0] > envelopes[j][0] && envelopes[i][1] > envelopes[j][1]) {
                max = Math.max(max, dp[j]);
            }
        }
        dp[i] = max + 1;
    }
    return dp[dp.length - 1];
    
};